crates/wasmtime/src/runtime/gc/enabled/structref.rs

```diff
@@ -0,0 +1,757 @@
+    pub fn field(&self, mut store: impl AsContextMut, index: usize) -> Result<Val> {
+        self._field(store.as_context_mut().0, index)
+    }
+
+    pub(crate) fn _field(&self, store: &mut StoreOpaque, index: usize) -> Result<Val> {
+        assert!(self.comes_from_same_store(store));
+
+        let mut fields = self._fields(store)?;
+        let len = fields.len();
+        fields
+            .nth(index)
+            .ok_or_else(|| anyhow!("cannot get field {index}: struct only has {len} fields"))
+    }
+
+    /// Set this struct's `index`th field.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error in the following scenarios:
+    ///
+    /// * When given a value of the wrong type, such as trying to set an `f32`
+    ///   field to an `i64` value.
+    ///
+    /// * When the field is not mutable.
+    ///
+    /// * When this struct does not have an `index`th field, i.e. `index` is out
+    ///   of bounds.
+    ///
+    /// * When `value` is a GC reference that has since been unrooted.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this reference is associated with a different store.
+    pub fn set_field(&self, mut store: impl AsContextMut, index: usize, value: Val) -> Result<()> {
+        self._set_field(store.as_context_mut().0, index, value)
+    }
+
+    pub(crate) fn _set_field(
+        &self,
+        store: &mut StoreOpaque,
+        index: usize,
+        value: Val,
+    ) -> Result<()> {
+        assert!(self.comes_from_same_store(store));
+
+        let mut store = AutoAssertNoGc::new(store);
+        let gc_ref = self.inner.try_gc_ref(&store)?.unchecked_copy();
+
+        let header = store.gc_store()?.header(&gc_ref);
+        debug_assert!(header.kind().matches(VMGcKind::StructRef));
+
+        let ty_index = header.ty().expect("structrefs should have concrete types");
+        let ty = StructType::from_shared_type_index(store.engine(), ty_index);
+
+        let layout = store
+            .engine()
+            .signatures()
+            .layout(ty_index)
+            .expect("struct types should have GC layouts");
+        let layout = match layout {
+            crate::vm::GcLayout::Array(_) => unreachable!(),
+            crate::vm::GcLayout::Struct(s) => s,
+        };
+
+        let structref = gc_ref.into_structref_unchecked();
+        debug_assert_eq!(ty.fields().len(), layout.fields.len());
+
+        let field_ty = match ty.field(index) {
+            Some(ty) => ty,
+            None => bail!(
+                "cannot set field {index}: struct only has {} fields",
+                ty.fields().len()
+            ),
+        };
+        ensure!(
+            field_ty.mutability().is_var(),
+            "cannot set field {index}: field is not mutable"
+        );
+        value
+            .ensure_matches_ty(&store, &field_ty.element_type().unpack())
+            .with_context(|| format!("cannot set field {index}: type mismatch"))?;
```

Comment:
I think that this is double-verifying the type because this type-check happens in `write_field` as well?